### PR TITLE
feat(external resources): MED-188 fix external resources

### PR DESCRIPTION
**Pull request by alicedraillard**
## Describe your changes
fixes:
- add and delete external resource favorites
- "Display all" on external resources

## Checklist tests
use module as usual

## Issue ticket number and link
[MED-188](https://jira.support-ent.fr/browse/MED-188)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [ ] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/openent/mediacentre/helper/ElasticSearchHelper.java [hunk 1]
```diff
@@ -33,48 +36,82 @@ private static String formatRegexp(String query) {
...
+                    if (sourceValue.equals(SourceEnum.SIGNET.method()))
+                        favorite.put(Field.ID, String.valueOf(favorite.getInteger(Field.ID)));
                 }
-                List<String> favorites = res.stream().map(f -> f.getString("id")).collect(Collectors.toList());
                 handler.handle(Future.succeededFuture(favorites));
             }
         });
     }
 
+    private static List<String> getFavoriteMatcher(String source) {
```
Comment (alicedraillard) [inline src/main/java/fr/openent/mediacentre/helper/ElasticSearchHelper.java:102]: returns fields to match to the favorite depending on the source

File: src/main/java/fr/openent/mediacentre/source/GAR.java [hunk 2]
```diff
@@ -160,10 +162,41 @@ public void plainTextSearch(String query, UserInfos user, Handler<Either<JsonObj
                 filteredResources.addAll(sortedMap.get(key));
             }
 
-            JsonObject response = new JsonObject()
-                    .put("source", GAR.class.getName())
-                    .put("resources", filteredResources);
-            handler.handle(new Either.Right<>(response));
+            favoriteService.get(SourceEnum.GAR.method(), user.getUserId())
```
Comment (alicedraillard) [inline src/main/java/fr/openent/mediacentre/source/GAR.java:165]: if a resource matches a favorite (depending on the fields structure_uai  id -must be defined-, and link) we set its fields:
favorite = true
favoriteId = the favorite's id

**Review by JinDouang — APPROVED** [review]

File: src/main/java/fr/openent/mediacentre/core/constants/SourceConstant.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package fr.openent.mediacentre.core.constants;
+
+public class SourceConstant {
+
+    public static final String SIGNET = "fr.openent.mediacentre.source.Signet";
```
Comment (aymericdevriesere) [inline src/main/java/fr/openent/mediacentre/core/constants/SourceConstant.java:5]: factoriser le .openent.mediacentre

File: src/main/java/fr/openent/mediacentre/helper/ElasticSearchHelper.java [hunk 4]
```diff
@@ -33,48 +36,82 @@ private static String formatRegexp(String query) {
...
+                    if (ar.failed()) {
+                        handler.handle(Future.failedFuture(ar.cause()));
+                    } else {
+                        List<JsonObject> resources = esFuture.result().stream()
+                                .filter(JsonObject.class::isInstance)
+                                .map(JsonObject.class::cast)
+                                .collect(Collectors.toList());
+                        List<JsonObject> favorites = favoriteFuture.result();
+                        List<String> favoriteMatcher = getFavoriteMatcher(source.getName());
+                        resources = resources
```
Comment (aymericdevriesere) [inline src/main/java/fr/openent/mediacentre/helper/ElasticSearchHelper.java:55]: Possible d'externaliser ? fonction dure à lire

File: src/main/java/fr/openent/mediacentre/source/GAR.java [hunk 5]
```diff
@@ -160,10 +162,41 @@ public void plainTextSearch(String query, UserInfos user, Handler<Either<JsonObj
...
+                    .onSuccess(favorites -> {
+                        List<JsonObject> formattedResources = filteredResources
```
Comment (aymericdevriesere) [inline src/main/java/fr/openent/mediacentre/source/GAR.java:167]: Si possible externaliser car fonction dure à lire

File: src/main/java/fr/openent/mediacentre/service/impl/DefaultFavoriteService.java [hunk 6]
```diff
@@ -42,6 +45,23 @@ public void get(String source, String userId, Handler<Either<String, JsonArray>>
...
+        Promise<JsonArray> promise = Promise.promise();
+
+        get(source, userId, event -> {
+            if (event.isRight()) {
+                promise.complete(event.right().getValue());
+            } else {
+                String message = String.format("[Mediacentre@%s::get] Error when fetching favorites: %s", this.getClass().getSimpleName(), event.left().getValue());
+                log.error(message);
+                promise.fail(ErrorEnum.ERROR_FAVORITE_FETCH.method());
+            }
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/mediacentre/service/impl/DefaultFavoriteService.java:59]: ta pas un FutureHelper ou un PromiseHelper pour rendre cette gestion générique ? comme ça tu t'embêterais moins à l'avenir ;)

File: src/main/resources/public/ts/controllers/home.ts [hunk 7]
```diff
@@ -179,13 +180,10 @@ class Controller implements IHomeViewModel {
...
-    seeMyExternalResource = (): void => {
-        let state = SEARCH_TYPE.PLAIN_TEXT;
-        let data: object = this.generateExternalResourceRequestBody();
+    seeMyExternalResource = async (): Promise<void> => {
+        let data: object = new PlainTextSearchData().build({query: ".*"}).toJson();
         this.$location.path(`/search/plain_text`);
-        this.$timeout(() => {
-            this.mainScope.$broadcast('search', {state, data, sources: [SOURCES.GAR]});
-        }, 300);
+        await this.mainScope.mc.startResearch(SEARCH_TYPE.PLAIN_TEXT, [SOURCES.GAR], data);
```
Comment (qpic-cgi) [inline src/main/resources/public/ts/controllers/home.ts:186]: alleluia !

**Review by qpic-cgi — APPROVED** [review]

File: src/main/java/fr/openent/mediacentre/enums/SourceEnum.java [hunk 8]
```diff
@@ -1,7 +1,10 @@
 package fr.openent.mediacentre.enums;
 
 public enum SourceEnum {
-    SIGNET("fr.openent.mediacentre.source.Signet");
+    SIGNET("fr.openent.mediacentre.source.Signet"),
```
[the remark this answers is not in the document]
Comment (alicedraillard) [reply src/main/java/fr/openent/mediacentre/enums/SourceEnum.java:4]: C'est mieux d'utiliser les enum mais j'ai du créer les constantes pour les mettre dans un switch/case

[the remark this answers is not in the document]
Comment (alicedraillard) [reply src/main/java/fr/openent/mediacentre/enums/SourceEnum.java:4]: ah oui je l'avais pas compris comme ca !

**Pull request merged**